### DISCUSSION TITLE
Configure more than one auth driver

### DISCUSSION THREAD

**Discussion by 9khaledSayed** [spatie/laravel-activitylog#834] [Q&A] [answered]
Is there is a way to use more than one auth driver I have two guards company and employee and want to store activities for both

'default_auth_driver' =>[ 'company', 'employee'],  trying to do this but not working

**Comment by Gummibeer** [comment]
@freekmurze what's the reason for this config value?
The auth facade/manager knows about its default and current guard. Would it be breaking to remove this custom logic and just use the auth guard provided by `Auth::guard()` which is the `auth.defaults.guard` or after successful authentication and `Auth::shouldUse()` call the one that's used to authenticate the request.

@9khaledSayed there's currently no way to use multiple guards directly. The only way I could imagine would be to don't do a single activity log before auth, call `Auth::shouldUse()` after authentication with the guard that was used (could be that Laravel does this for you) and set the package config value to `null` as this will use the `Auth::defaultGuard()` on service initialization.

**Reply by 9khaledSayed** [reply]
thanks for replying
```php
'default_auth_driver' => null
```
my problem is when i'm authenticated as a company or employee and do an activity the causer id is set to null

this is my guards
```php
'guards' => [
        'web' => [
            'driver' => 'session',
            'provider' => 'users',
        ],
        'company' => [
            'driver' => 'session',
            'provider' => 'companies',
        ],
        'employee' => [
            'driver' => 'session',
            'provider' => 'employees',
        ],

        'api' => [
            'driver' => 'token',
            'provider' => 'employees',
            'hash' => false,
        ],
    ],
```

**Reply by Gummibeer** [reply]
Could you do some debugging for me in your app?

`dump('authDriver', $this->authDriver);` after this line
https://github.com/spatie/laravel-activitylog/blob/ade270f291f4cb5883b3653919304a0b4e1cc284/src/ActivityLogger.php#L37

and `dd('activity', $this->activity);` after this one but inside the `if`
https://github.com/spatie/laravel-activitylog/blob/ade270f291f4cb5883b3653919304a0b4e1cc284/src/ActivityLogger.php#L232

After adding this in your `vendor` files you will have to login as anything and enforce an activity log.

In theory the result should be `'company', Company(...)` and `'employee', Employee(...)` - depending on your used auth guard.

**Reply by 9khaledSayed** [reply]
`dump('authDriver', $this->authDriver);` can you explain to me " $this  " refer to which class so i can debugging

Note that these two lines are already exist

`$this->authDriver = $config['activitylog']['default_auth_driver'] ?? $auth->getDefaultDriver(); `

`->causedBy($this->auth->guard($this->authDriver)->user()); `

**Reply by Gummibeer** [reply]
I meant that you should copy exactly these two debug statements at the given position.^^
To debug the auth driver/guard used by the activity logger and which model is assigned as causer to the activity.

**Reply by 9khaledSayed** [reply]
thank you so much for helping me my problem is solved when i do that 
 $this->authDriver = Auth::guard('company')->check()? 'company' : 'employee';

**Reply by Gummibeer** [reply]
That's the dirty fix. Just to be sure: don't do this in `vendor`. Instead create a class that extends the original, override the construct how you need it and rebind it in your app service provider to the container.
Otherwise your changes will be lost after next update and you will go crazy fix it again.^^

**Reply by 9khaledSayed** [reply]
thank you for this information ok i will create an external class and override the __construct
![image](https://user-images.githubusercontent.com/40536997/104362945-23ebca80-551d-11eb-938e-502f7ac1d754.png)

but now how can i bind it in app service provider

**Comment by Gummibeer** [comment, marked as answer]
Hey @9khaledSayed ,
so we can solve this discussion I will post these snippets as a "real" answer as the comments can't be flagged as accepted answer.
If you use these snippets everything should work. The first is the whole file the second is one line you have to copy in the surrounding method. 😉

**app/ActivityLog/ActivityLogger.php**
```php
<?php

namespace App\ActivityLog;

use Illuminate\Auth\AuthManager;
use Illuminate\Contracts\Config\Repository as Config;
use Spatie\Activitylog\ActivityLogger as SpatieActivityLogger;
use Spatie\Activitylog\ActivityLogStatus;

class ActivityLogger extends SpatieActivityLogger;
{
    public function __construct(AuthManager $auth, Config $config, ActivityLogStatus $logStatus)
    {
        parent::__construct($auth, $config, $logStatus);

        $this->authDriver = $auth->guard('company')->check() ? 'company' : 'employee'; // this can be customized to any logic you want/need
    }
}
```

**app/Providers/AppServiceProvider.php**
```php
use Spatie\Activitylog\ActivityLogger as SpatieActivityLogger;
use App\ActivityLog\ActivityLogger;

// ...

public function register(): void
{
    // ...
    $this->app->bind(SpatieActivityLogger::class, ActivityLogger::class);
    // ...
}

// ...
```

**Reply by 9khaledSayed** [reply]
Thank you so much bro

**Reply by Gummibeer** [reply]
No Problem. Happy to help.
Send us a postcard if you haven't already. 😉💌📯
https://spatie.be/open-source/postcards

**Reply by heyder777** [reply]
Thank you. But it gave me an error. That's why I changed.
![image](https://github.com/spatie/laravel-activitylog/assets/28921795/c3cbda92-618f-4261-b097-57485c73d2b5)
